Esp32 can driver

nos86 opened the pull request:
This PR adds CAN Device added for ESP32.
see issue #5087

Commit:
first commit

nos86 committed:
Enabled module

nos86 committed:
Stable basic code for module loading in Python

nos86 committed:
Full initialization and dummy write

nos86 committed:
added function to retrieve information of CAN STATUS

nos86 committed:
Function SEND and RECV implemented

nos86 committed:
Update quickref.rst

Added quick commands for CAN usage

nos86 committed:
improve object print

nos86 committed:
remove prefix CAN_ from constants

nos86 committed:
implementation of hardware filter

nos86 committed:
update includePath

nos86 commented:
> The command "ls -l ports/minimal/build/firmware.bin" exited with 0.
0.01s$ tools/check_code_size.sh
Old size: 68848 new size: 68852
Validation failure: Core code size increased
The command "tools/check_code_size.sh" exited with 1.

Full log at https://travis-ci.org/micropython/micropython/jobs/609232858?utm_medium=notification&utm_source=github_status

I do not expect any increase in basic firmware. All changes are applied to ports/esp32.
What I have to check to pass this test?

camosoul commented:
I'm a noob, so I can't help... Except to maybe draw some attention to this. It sets off the 'wut' meter that MicroPython made a 1,.0 release without this... **This is in desperate need.**

Tbruno25 commented:
I agree. Would love to interface with CAN with the esp32.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_rx_waiting_messages), MP_ROM_PTR(&machine_hw_can_get_rx_waiting_messages_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_state), MP_ROM_PTR(&machine_hw_can_get_state_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_clear_tx_queue), MP_ROM_PTR(&machine_hw_can_clear_tx_queue_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_clear_tx_queue), MP_ROM_PTR(&machine_hw_can_clear_rx_queue_obj)},
```
Last one should be clear_rx_queue.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_rec), MP_ROM_PTR(&machine_hw_can_get_rec_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_tx_waiting_messages), MP_ROM_PTR(&machine_hw_can_get_tx_waiting_messages_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_rx_waiting_messages), MP_ROM_PTR(&machine_hw_can_get_rx_waiting_messages_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_state), MP_ROM_PTR(&machine_hw_can_get_state_obj)},
```
This is just `state` in STM32.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+    { MP_ROM_QSTR(MP_QSTR_deinit), MP_ROM_PTR(&machine_hw_can_deinit_obj) },
+    { MP_ROM_QSTR(MP_QSTR_send), MP_ROM_PTR(&machine_hw_can_send_obj)},
+    { MP_ROM_QSTR(MP_QSTR_recv), MP_ROM_PTR(&machine_hw_can_recv_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_get_tec), MP_ROM_PTR(&machine_hw_can_get_tec_obj)},
```
These four methods don't exist in STM32 -- instead it just has `any()`.

If your application needs these counters, could it possibly be collapsed to a single method `stats` returning a tuple or something?

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+    { MP_OBJ_NEW_QSTR(MP_QSTR_clear_tx_queue), MP_ROM_PTR(&machine_hw_can_clear_tx_queue_obj)},
+    { MP_OBJ_NEW_QSTR(MP_QSTR_clear_tx_queue), MP_ROM_PTR(&machine_hw_can_clear_rx_queue_obj)},
+    //CAN_MODE
+    { MP_ROM_QSTR(MP_QSTR_MODE_NORMAL), MP_ROM_INT(CAN_MODE_NORMAL) },
```
How do these map to the STM32 modes? NORMAL / LOOPBACK / SILENT / SILENT_LOOPBACK?

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+    { MP_ROM_QSTR(MP_QSTR_MODE_NO_ACK), MP_ROM_INT(CAN_MODE_NO_ACK) },
+    { MP_ROM_QSTR(MP_QSTR_MODE_LISTEN_ONLY), MP_ROM_INT(CAN_MODE_LISTEN_ONLY) },
+    //CAN_STATE
+    { MP_ROM_QSTR(MP_QSTR_STATE_STOPPED), MP_ROM_INT(CAN_STATE_STOPPED) },
```
Same for the STM32 states? STOPPED / ERROR_ACTIVE / ERROR_WARNING / ERROR_PASSIVE / BUS_OFF.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+STATIC MP_DEFINE_CONST_DICT(machine_can_locals_dict, machine_can_locals_dict_table);
+
+//Python object definition
+const mp_obj_type_t machine_can_type = {
```
I think this can move to machine_can.c -- a definition shouldn't be in a header.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+STATIC mp_obj_t machine_hw_can_clear_tx_queue(mp_obj_t self_in);
+STATIC mp_obj_t machine_hw_can_clear_rx_queue(mp_obj_t self_in);
+
+//Python function declarations
```
These don't need to be in the header file -- for consistency move to below the function in the .c file they're referencing.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+} machine_can_obj_t;
+
+//Internal functions
+STATIC can_state_t _machine_hw_can_get_state();
```
STATIC methods shouldn't be in the header file -- the point of STATIC is that they're private to the C file. So if you need the forward declarations, just put them in the .c file.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+
+#include "py/obj.h"
+
+#define DEVICE_NAME "CAN"
```
Anything that's not needed outside machine_can.c can move into the .c file.

I think that applies to all of these #defines, and the two structs below.

jimmo commented on ports/esp32/machine_can.h:
```diff
@@ -0,0 +1,137 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014 Damien P. George
```
Up to you if you want to add your name, but at least update the year.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,336 @@
+
+#include <machine_can.h>
+
+//#if MICROPY_HW_ENABLE_CAN
```
This should be uncommented

jimmo commented on .vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Please remove this file.

jimmo commented:
> I do not expect any increase in basic firmware. All changes are applied to ports/esp32.
What I have to check to pass this test?

hi @nos86 thanks for sending this! I know a lot of people are keen to see CAN support on ESP32.

I've made some basic comments on the code, but you'll also need to do some minor coding convention fixes too (things like newline at EOF, spaces around operators, spaces after `//` in comments, etc). Just to keep consistency with the rest of the code base.

But the bigger question here is compatibility with the STM32 CAN driver. Ideally they would be identical APIs (and we'd move pyb.CAN into machine.CAN on the STM32 port) however I know that there needs to be some compromise and the STM32 version will need to change when it's moved to machine.CAN. But as the first implementation of machine.CAN this PR is kind of setting a precedent for what the API will look like. I've posted some simple questions inline in the code, but I don't know much about CAN so perhaps you or others have some thoughts on how to proceed.

> I'm a noob, so I can't help... Except to maybe draw some attention to this. It sets off the 'wut' meter that MicroPython made a 1.0 release without this... This is in desperate need. Why it isn't a high priority for the uPython peeps, I don't understand.

@camosoul I'm sure you're aware that the ESP32 port came a long time after MicroPython 1.0 and that everyone's priorities are different. Like I said above though, it's great to see this PR.

mattytrentini commented:
Hey @chrismas9 do you have any opinions on this PR? I recall that you've worked with CAN quite a lot...
For reference, the STM32 implementation is currently in [pyb.CAN](https://docs.micropython.org/en/latest/library/pyb.CAN.html).

camosoul commented:
> > I'm a noob, so I can't help... Except to maybe draw some attention to this. It sets off the 'wut' meter that MicroPython made a 1.0 release without this... This is in desperate need. Why it isn't a high priority for the uPython peeps, I don't understand.
> 
> @camosoul I'm sure you're aware that the ESP32 port came a long time after MicroPython 1.0 and that everyone's priorities are different. Like I said above though, it's great to see this PR.

Nope, didn't realize that. Makes sense. I appreciate being rid of (a portion of) my derpage.

nos86 commented:
Hello @jimmo,
I'll work on your comments in the incoming days in order to align the code with your expectation.
In a general way, I'll do a cross check between STM32 and ESP32 in order to guaratee the harmonization between method signatures.
In any case, please consider that this code is related to ESP32 hardware, so in same case the complete harmonization will be not possible (e.g. CAN Modes is the same defined in the ESP32 framework).

nos86 committed:
Changed author

nos86 committed:
STM32 Harmonization: rename get_state() to state(), add info() and any() functions and remove functions now included in info()

nos86 committed:
STM32 Harmonization: added restart() signature

nos86 committed:
STM32 Harmonization: move private methods in c file

nos86 committed:
Enabled compiler switch for module

nos86 committed:
Code Cleaning: Reordered the function sequence

nos86 committed:
STM32 Harmonization: init() returns mp_const_none and raise exceptions instead Error Logs

jimmo commented:
Great, thanks for the updates. Let me know when you've finished and I'll take another look.

> In any case, please consider that this code is related to ESP32 hardware, so in same case the complete harmonization will be not possible (e.g. CAN Modes is the same defined in the ESP32 framework).

Yes, however it's still the same underlying CAN protocol. I understand that the STM32 driver might need to change too, but wherever possible the API should be built around the fundamental CAN states and operations rather than the ESP32-specific terminology provided by the IDF (or the STM32 HAL). Of course there are exceptions when ESP32 supports additional CAN functionality that STM32 doesn't (and vise versa), but where the functionality overlaps the API should be the same.

nos86 committed:
STM32 Harmonization: Alignment of init function to pyCan.

nos86 committed:
STM32 Harmonization: Alignment of send(...) and recv(...) methods to pyBoard signature and return values

nos86 committed:
removed auto_restart option since the esp32 framework doesn't have this feature. It could be implemented in the feature if needed

nos86 committed:
Added signature for RX Callback; currently this method is not implemented

nos86 committed:
Python Interface has been moved from header to  source

nos86 committed:
Removed redundant function

nos86 committed:
beautify

nos86 committed:
STM32 Harmonization: implemented same CAN_MODE

nos86 commented:
I'm working on that. Next week I'll do some tests on a physical device connected to a CAN bus.
I cannot promise I'm able to finish by end of the year.

IveJ commented:
Hi,

Refer link below and attached file for your reference.
  report-tfm-eduard-valentino.pdf
<https://drive.google.com/file/d/0B_8OJvgHabFGbnJKbm8wOWZTeFZsNUswUThMZTNvWkstLV9z/view?usp=drivesdk>



On Mon, Dec 16, 2019, 19:07 Integr@te System <datacentralize@gmail.com>
wrote:

> Hi,
>
> Some docs you can refer to you project:
>
>
> https://www.hackster.io/news/the-can32-an-esp32-based-can-bus-board-af5c9e1bd8ec
>
>
> https://docs.micropython.org/en/latest/library/pyb.CAN.html
>
> On Mon, Dec 16, 2019, 02:48 Salvo Musumeci <notifications@github.com>
> wrote:
>
>> I'm working on that. Next week I'll do some tests on a physical device
>> connected to a CAN bus.
>> I cannot promise I'm able to finish by end of the year.
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/micropython/micropython/pull/5310?email_source=notifications&email_token=AEYAML6APZQFSZTESECWSULQY2CZNA5CNFSM4JKXHE2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG5AVAQ#issuecomment-565840514>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEYAML2ZJS6JG6UTJYXDBYTQY2CZNANCNFSM4JKXHE2A>
>> .
>>
>
  CAN_Theory&Case_Study.pdf
<https://drive.google.com/file/d/0B_8OJvgHabFGQTQwZkRJRnNMOUplbnY3bDZUeTVrNlNnamtj/view?usp=drivesdk>

  canprimer_v2.pdf
<https://drive.google.com/file/d/0B_8OJvgHabFGU01HYzN3Njd4TlJPeEI0dk5Iem5Ga0U0b0Y4/view?usp=drivesdk>

nos86 committed:
minor fix

nos86 committed:
manage flags for loopback and extframe

nos86 committed:
removed unused input variable from any()

nos86 committed:
bugfix

nos86 committed:
Improved print function

nos86 committed:
Implementation of restart function

nos86 committed:
data type bugfix

nos86 committed:
Added alert function

nos86 committed:
Added docs for ESP32 CAN

nos86 committed:
Removed filter setup from init.

nos86 committed:
STM32 Harmonization

nos86 committed:
Integrated you docs with existing one

Commit:
Code cleaning

Commit:
WIP: first implementation of filters

Commit:
WIP

nos86 committed:
Syntax Correction

nos86 committed:
CAN configuration is stored in object struct

nos86 committed:
Bugfixes

nos86 committed:
WIP: implementation of setfilter()

nos86 committed:
Completed setfilter()

nos86 committed:
added clearfilter()

nos86 committed:
Added unittest as example

nos86 committed:
removed machine_can.h

nos86 committed:
Bugfix

nos86 committed:
Docs updated

nos86 committed:
Update Author name

nos86 committed:
remove .vscode folder

nos86 commented:
I've finished the implementation..!
Now the Python interface similar to pyCAN as much as possible.
Some specific feature of STM32 that are not available on esp32 has been skipped. In any case I've tried to keep the arguments and return response close to STM32 implementation.

I have added some documentation as well and fixed almost all issues you highlighted.
Moreover I added a small py code under example that can be used as unittest. I don't know if it is the right place.

Now, the implementation is stable and I don't expect any breaking changes in the future.
For sure, some beautify has to be applied, but for me this is time to take some vacation.. :)

I really appreciate if someone can test the implementation during Christmas period.

Merry Christmas and Happy New Year.

camosoul commented:
At what point will this magnificent creation be included in the daily builds?

jimmo commented:
I'm going to look into this in more detail shortly, but just a few quick notes. Overall though, it's looking really good. Thank you!

 - This PR adds machine.CAN but the documentation seems to be describing esp32.CAN.  I think this makes sense to go in machine, so please update the docs.
 - There's still some minor formatting issues -- spaces around brackets and operators and stuff.
 - Similarly, please PEP8 the example code. As this is a generic API, it doesn't need the esp32 prefix -- perhaps create examples/machine/can.py
 - We might need to think about cutting down the number of constants exposed, that's a lot of QSTR data (which affects everyone even if they don't use CAN). What we've done in other places (e.g. BLE) is to provide sample code in the docs that contain the list of constants. (See the lists of consts() at http://docs.micropython.org/en/latest/library/ubluetooth.html#ubluetooth.BLE.irq )
 - For now please remove the unsupported parts (unless you plan to add them in before getting this merged).
 - It would be really good if you could comment the ESP32-only parts. I notice there's already some of this in the machine_can_locals_dict_table, but also the kwargs to init and setfilter, etc.

julianpass commented:
@nos86 I was able to compile and run your changes. I also was able to send and receive messages via CAN. In the next weeks, I will start to implement my OTA via CAN project which will update a Teensy via CAN with hex files uploaded onto the ESP32. I will keep you updated if I find any issues. Thank you very much for enabling CAN with Micropython on ESP32!

nos86 commented:
@jimmo, some comments from my side

> This PR adds machine.CAN but the documentation seems to be describing esp32.CAN. I think this makes sense to go in machine, so please update the docs.

The implementation of CAN is close to pyb implementation but is not the same. 
From one hand, I improved some functionality and from the other hand, the different hardware is reflected in the implementation.

> There's still some minor formatting issues -- spaces around brackets and operators and stuff.

Do you have a standard to refer to?

nos86 committed:
Formatting sources

nos86 committed:
Removed additional constants and added in documentation

nos86 committed:
Documentation bugfix

nos86 committed:
removed implemented methods

nos86 commented:
For anyone interested in a nightly built firmware: https://github.com/nos86/micropython/releases/download/esp32-can-driver-v2/firmware.bin

jimmo commented:
> @jimmo, some comments from my side
> 
> > This PR adds machine.CAN but the documentation seems to be describing esp32.CAN. I think this makes sense to go in machine, so please update the docs.
> 
> The implementation of CAN is close to pyb implementation but is not the same.
> From one hand, I improved some functionality and from the other hand, the different hardware is reflected in the implementation.

Sounds good. I think the overarching point here is that despite the hardware differences, there's at least some minimal functionality that both ports support, and so Python code that only uses that minimal functionality should work on both ESP32 and STM32 unchanged.

This might require that in the process of moving pyb.CAN to machine.CAN on the STM32 some things will change, but it would be good if the code could make it clear "this part is ESP32 specific" vs "this part should be the same on all ports".

But my comment was more that the PR implements machine.CAN, while the docs in this PR are for esp32.CAN.  (I think it should be machine.CAN though).

> 
> > There's still some minor formatting issues -- spaces around brackets and operators and stuff.
> 
> Do you have a standard to refer to?

https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md#c-code-conventions is probably the closest thing.

Also please remove the .gitignore changes.

camosoul commented:
I am so twitchy now...

nos86 committed:
beautify: brackets inline

nos86 committed:
documentation update: from esp32 specifc module to machine module

nos86 committed:
revert unnecessary changes

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+        ARG_timeout
+    };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_list, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_PTR(&mp_const_none_obj)}},
```
`.u_rom_obj = MP_ROM_NONE`

And spaces around the `{` and `}`. e.g.  `{ MP_QSTR_list, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} }` seems to be the convention.

nos86 replied:
compiler cannot resold MP_ROM_NONE, I have to add any further header file?

nos86 replied:
I think this definition is not available since my branch started before this addition. Actually I don't know I rebase the branch without making a mess.

jimmo replied:
The branch should rebase cleanly (I checked). You'll need to do this anyway, and clean up the commits a bit in the process. Probably the best way to do this is with an interactive rebase.

```
git fetch origin
git rebase -i origin/master
```

You'll be prompted to select what you want to do with each of the commits. I suggest using "fix" to combine commits together (and possibly reorder them). And then possibly run the `rebase -i` again and go through and use `reword` to update the commit messages.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+#define CAN_BAUDRATE_1M 1000
+
+#define ESP_STATUS_CHECK(status)   \
+    if (status != ESP_OK)          \
```
Open brace on same line as `if`

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+    {                              \
+        mp_raise_OSError(-status); \
+    }
+typedef enum _filter_mode_t{
```
Always a space before `{`, i.e.

`if (x) {` or `typedef enum foo {`

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+        ARG_self
+    };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_data, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL}},
```
No default required for required params.

tve replied:
Why was this marked as resolved?

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+STATIC machine_can_obj_t machine_can_obj = {{&machine_can_type}, .config = &can_config};
+
+// INTERNAL FUNCTION Return status information
+STATIC can_status_info_t _machine_hw_can_get_status(){
```
There isn't really a convention for leading-underscore naming for internal stuff. And the comment probably isn't needed -- making it static is a clear marker of what's going on.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+    can_status_info_t status;
+    uint32_t err_code = can_get_status_info(&status);
+    if (err_code != ESP_OK)
+    {
```
Open brace goes on previous line.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+        addr = (addr & 0x1FFFFFFF) << 3 | (rtr ? 0x04 : 0);
+        mask = (mask & 0x1FFFFFFF) << 3 | 0x03;
+        preserve_mask = 0;
+    }else if (self->config->filter.single_filter){
```
Space before `else`

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+    list->items[4] = MP_OBJ_NEW_SMALL_INT(self->num_bus_off);
+    list->items[5] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_tx);
+    list->items[6] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_rx);
+    list->items[7] = mp_const_none;
```
What's this for? I see the documentation just says "Reserved"

nos86 replied:
Current ESP-IDF implementation for CAN device do not foresee events management. So in order to have a cross-compatibility with STM32 I left the same array order information, but for those information not available I indicated RESERVED in the documentation. Maybe for the next implementation step, I'll introduce some free RTOS task to monitor the status of device and give the information to the mp layer.

jimmo replied:
Ah yes, sorry I missed this in the STM32 version. Thanks

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+    if (length > 8){
+        mp_raise_ValueError("CAN data field too long");
+    }
+    uint8_t flags = (args[ARG_rtr].u_bool == true ? CAN_MSG_FLAG_RTR : CAN_MSG_FLAG_NONE);
```
`== true` not required

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,673 @@
+        memcpy(mv->items, rx_message.data, rx_message.data_length_code);
+    }
+    items[0] = MP_OBJ_NEW_SMALL_INT(rx_message.identifier);
+    items[1] = rx_message.flags && CAN_MSG_FLAG_RTR > 0 ? mp_const_true : mp_const_false;
```
mp_obj_new_bool

jimmo reviewed (commented):
Thanks for making the fixes! I've added some more comments. Sorry for the nitpicks, it's mostly just style stuff, but consistency is important.

Other than that, I think this is looking pretty good. Need someone with more CAN + STM32 experience to comment on whether this is going to be feasible to implement the same API on STM32.

nos86 committed:
code beautify

nos86 commented:
Hello @jimmo,
I have reviewed my code. Unfortunately the missing availability of an automatic beautifier on VSCode makes the cosmetic aspects so hard and time consuming.

Regarding STM32 harmonization, I think the point is complex and it should be discussed together with the developer of STM32 CAN device.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -102,50 +101,47 @@ STATIC mp_obj_t machine_hw_can_init_helper(machine_can_obj_t *self, size_t n_arg
 STATIC void _machine_hw_can_set_filter(machine_can_obj_t *self, uint32_t addr, uint32_t mask, uint8_t bank, bool rtr){
     //Check if bank is allowed
-    if (bank < 0 && bank > ((self->extframe && self->config->filter.single_filter) ? 0 : 1)){
+    if ( bank < 0 && bank > ( (self->extframe && self->config->filter.single_filter) ? 0 : 1 ) ){
```
Space around parens not required here. The changes I was suggesting only applied to braces. Please undo these changes.

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -102,50 +101,47 @@ STATIC mp_obj_t machine_hw_can_init_helper(machine_can_obj_t *self, size_t n_arg
     uint32_t err_code = can_get_status_info(&status);
-    if (err_code != ESP_OK)
-    {
+    if (err_code != ESP_OK){
```
Missing space before {

jimmo commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,669 @@
+        memcpy(mv->items, rx_message.data, rx_message.data_length_code);
+    }
+    items[0] = MP_OBJ_NEW_SMALL_INT(rx_message.identifier);
+    items[1] = mp_obj_new_bool(rx_message.flags && CAN_MSG_FLAG_RTR > 0 ? true : false);
```
the `? true : false` is unnecessary.

jimmo commented:
> Hello @jimmo,
> I have reviewed my code. Unfortunately the missing availability of an automatic beautifier on VSCode makes the cosmetic aspects so hard and time consuming.

I agree it would be really good if there was an automated tool for this and it has been discussed before. You might want to take a look at astyle or clang-format (I'm not sure if there's an existing configuration for either, but see https://github.com/micropython/micropython/issues/4223 for some notes).

But for any given thing, there'll be a line somewhere else in the codebase you can copy the style from. And you can do most of these fixups with find/replace, i.e. `){` --> `) {`.  `( (` --> `((`.

Just a note on the rebase / commit squashing. The idea is to end up with it being a small number of commits where each does a single thing with a clear description.

e.g.
```
esp32: Add machine.CAN driver.
docs: Add docs for machine.CAN.
```

jimmo commented:
FWIW -- here's a sample `astyle` invocation that does a pretty good job of taking care of the remaining issues:

```
astyle --style=google --suffix=none --indent=spaces=4 --attach-closing-while --add-brackets --indent-switches --indent-after-parens --indent-continuation=1 --indent-labels --indent-preproc-define --indent-preproc-cond --pad-header --unpad-paren --add-braces --convert-tabs --keep-one-line-statements --max-code-length=200 ports/esp32/machine_can.c
```

nos86 committed:
Beautify

nos86 commented:
Hi @jimmo,
thank you very much for the beautifier hint. I have applied it and reviewed other open points.
Can you review it and give me a feedback?
If everything is ok, I'll proceed to rebase the code closing the last open points and grouping all commits in one.

Sashkoiv commented:
> For anyone interested in a nightly built firmware: https://github.com/nos86/micropython/releases/download/esp32-can-driver-v2/firmware.bin

I'm getting this:
```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 
ets Jun  8 2016 00:22:57
```
Flashed using:
```
esptool.py --port /dev/ttyUSB0 erase_flash
esptool.py --port /dev/ttyUSB0 --baud 115200 write_flash --flash_size=detect 0x0 ./esp/firmwareESP32CAN.bin
picocom /dev/ttyUSB0 -b115200
```

camosoul commented:
...so?

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+        addr = ( ( (addr & 0x7FF) << 5 ) | (rtr ? 0x10 : 0) );
+        mask = ( (mask & 0x7FF) << 5 );
+        preserve_mask = 0xFFFF << ( bank == 0 ? 16 : 0 );
+        if ( (self->config->filter.acceptance_mask & preserve_mask) == ( 0xFFFF << (bank == 0 ? 16 : 0) ) ) {
```
why not `(self->config->filter.acceptance_mask & preserve_mask) == preserve_mask` ?

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+    } else {
+        addr = ( ( (addr & 0x7FF) << 5 ) | (rtr ? 0x10 : 0) );
+        mask = ( (mask & 0x7FF) << 5 );
+        preserve_mask = 0xFFFF << ( bank == 0 ? 16 : 0 );
```
`preserve_mask = bank == 0 ? 0xffff0000 : 0xffff;`

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+            preserve_mask = 0;
+        } else {
+            addr = addr << (bank == 1 ? 16 : 0);
+            mask = mask << (bank == 1 ? 16 : 0);
```
```
if (bank == 1) {
  addr <<= 16;
  mask <<= 16;
}
```

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+    can_status_info_t status;
+    uint32_t err_code = can_get_status_info(&status);
+    if (err_code != ESP_OK) {
+        mp_raise_OSError(-err_code);
```
Should use `check_esp_err()` everywhere.

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+    if (status != ESP_OK) {
+        mp_raise_OSError(-status);
+    }
+    mp_hal_delay_ms(200);
```
Whoa! That's a long delay! Why? Or how can it be avoided?

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+    byte rx_state;
+    bool extframe : 1;
+    bool loopback : 1;
+    uint16_t num_error_warning; //FIXME: populate this value somewhere
```
If these values are not being populated, then maybe leave them out?

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+
+// Clear TX Queue
+STATIC mp_obj_t machine_hw_can_clear_tx_queue(mp_obj_t self_in) {
+    return mp_obj_new_bool(can_clear_transmit_queue() == ESP_OK);
```
Why return a boolean instead of raising an error that provides info about the failure?

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+#define ESP_STATUS_CHECK(status)   \
+    if (status != ESP_OK) {        \
+        mp_raise_OSError(-status); \
+    }
```
Please use `check_esp_err()`

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+STATIC mp_obj_t machine_hw_can_init(size_t n_args, const mp_obj_t *args, mp_map_t *kw_args) {
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    if (self->config->initialized) {
+        ESP_LOGW(DEVICE_NAME, "Device is already initialized");
```
Why does this not raise an error? MP doesn't use ESP_LOGW in general...

IhorNehrutsa replied:
@twe @dpgeorge @jimmo 
I find it difficult to choose the type of error in cases
"Device is already initialized" and "Device is not initialized"
I can raise
_mp_raise_ValueError(msg);_
_mp_raise_TypeError(msg);_
_mp_raise_NotImplementedError(msg);_
_mp_raise_OSError(errno);_
but I don't like them.
In fact, it is an algorithm user program bug.
Twice init, twice deinit etc.

I want to help a user change his program.
I prefer to raise 
mp_raise_OSError(EPERM, "Device is already initialized")  
 with console output like:
**>>>** OSError: [Errno 1] EPERM: Device is already initialized

tve commented on ports/esp32/machine_can.c:
```diff
@@ -0,0 +1,672 @@
+        if (self->config->initialized) {
+            // The caller is requesting a reconfiguration of the hardware
+            // this can only be done if the hardware is in init mode
+            ESP_LOGW(DEVICE_NAME, "Device is going to be reconfigured");
```
MP doesn't use the ESP IDF logging facility.

tve reviewed (commented):
This is an interesting PR, thank you! I like that it comes with docs and an example :-) !
Things I would like to see improved:
- make it truly compatible with the stm32 CAN driver, not mostly. I don't mean exact feature parity, but not having different arg lists, etc.
- make the docs generic and point out esp specific stuff as notes
- use check_esp_err to raise OS errors

hklang10 commented:
For a personal project, I have made some updates/small changes to the excellent work by @nos86 

The work is currently in my branch https://github.com/hklang10/micropython/tree/esp32-can-updates

Current status:
* Has been updated to build against the latest micropython/master
* has some documentation updates
* Error message handling has been improved
* implement a timeout for the recv() method

There are some things to resolve with init(), the get_alterts() method needs to be implemented.

For me, this is building OK with espIDF 3.3.2

Comments/suggestions welcome.

kdschlosser commented:
I am guessing that this went stale. 👎

mattytrentini commented:
> I am guessing that this went stale. 👎

Maybe someone can pick it up? I'd give it a shot but don't have any ESP32 boards with CAN - or any CAN peripherals...

kdschlosser commented:
There is an Adafruit circuit python mcp2515 CAN library that was just made a few months ago. It appears to not be finished tho I am not sure and I have not tested it as of yet.

I didn't even know that the ESP32 had a built in CAN interface until I came across this PR. I was planning on using an external interface. I have not been able to locate a complete library\module for an external interface. 

I could probably port one of the C/C++ libraries over to python pretty easily so it would be able to run using MicroPython, my concern with doing something like that would be the speed reduction compared to running it in C code. I am not sure if the speed reduction would be an amount that could cause frames to get missed.

I know enough C/CPP code to get me into trouble, defiantly not enough knowledge to complete this kind of a project. It does appear that this PR is mostly complete with only a few small changes needed.

IhorNehrutsa commented:
to @nos86: Dear Salvo, do you have the strength and time to continue working on this PR? Thank you very much for you work.

nos86 commented:
Hello @IhorNehrutsa,
unfortunately, I have no time for the moment and also the man project that asked me to develop this functionality on Micropython has been concluded.

Basically, my code is completely working and implements all basic functions.
The major problems are related to the harmonization and cosmetic aspects.

I made two loops of harmonization but it seems that it's not enough yet.

If I remember well the esp32-can-driver-harmonization is born from the esp32-can-driver and it is created for the harmonization process with existing STM32 firmware. Please continue from it.

jimmo commented:
Closing in favour of #7381 -- thanks for your work on this @nos86, and thanks @IhorNehrutsa for continuing it.

FWIW, this PR was the original motivation for the work to get automatic code formatting implemented, and the process for submitting PRs has improved dramatically, avoiding the frustration of dealing with style/formatting nit picking.
